File: src/PatternNoWarn.fs
module PatternNoWarn

open SkyrimPlatform
open BindingWrapper

#nowarn "25"

open Core
open FSharp.UMX
let (|LessHealth|MoreHealth|EqualHealth|) (ctxHealth: HealthContext) =
  
  let delta = System.Math.Abs(%ctxHealth.CurrentHealth - %ctxHealth.LastHealth)
      
  match ctxHealth.CurrentHealth with
  | _ when %ctxHealth.CurrentHealth > %ctxHealth.LastHealth -> MoreHealth delta
  | _ when %ctxHealth.CurrentHealth = %ctxHealth.LastHealth -> EqualHealth
  | _ when %ctxHealth.CurrentHealth < %ctxHealth.LastHealth -> LessHealth delta
    
let (|PercentLessHealth|PercentMoreHealth|PercentEqualHealth|) (ctxHealth: HealthContext) =

  let percentEquality =
    if System.Math.Abs(%ctxHealth.CurrentHealthPercent - %ctxHealth.LastHealthPercent) > 0.02 then
      false
    else
      true
        
  match percentEquality with
  | true -> PercentEqualHealth
  | _ when %ctxHealth.CurrentHealthPercent > %ctxHealth.LastHealthPercent -> PercentMoreHealth
  | _ when %ctxHealth.CurrentHealthPercent < %ctxHealth.LastHealthPercent -> PercentLessHealth
  
let (|IsManaShieldEffect|_|) (equal, objRef: ObjectReference) =
  if equal then
    match sp.Actor.From (objRef :> Form) with
    | Some actor -> Some actor
    | _ -> None
  else None
  
let (|MagicEffectFromForm|_|) (someForm: Form option) =
  match someForm with
  | Some form ->
    match sp.MagicEffect.From(form) with
    | Some me -> Some me
    | _ -> None
  | _ -> None
  
let (|PerkFromForm|_|) (someForm: Form option) =
  match someForm with
  | Some form ->
    match sp.Perk.From(form) with
    | Some p -> Some p
    | _ -> None
  | _ -> None

File: jsc/PatternNoWarn.js
import { equals, compare } from "./fable_modules/fable-library.3.6.2/Util.js";
import { FSharpChoice$3 } from "./fable_modules/fable-library.3.6.2/Choice.js";
import * as skyrimPlatform from "../src/skyrimPlatform.declare";

export function $007CLessHealth$007CMoreHealth$007CEqualHealth$007C(ctxHealth) {
    const delta = Math.abs(ctxHealth.CurrentHealth - ctxHealth.LastHealth);
    if (compare(ctxHealth.CurrentHealth, ctxHealth.LastHealth) > 0) {
        return new FSharpChoice$3(1, delta);
    }
    else if (equals(ctxHealth.CurrentHealth, ctxHealth.LastHealth)) {
        return new FSharpChoice$3(2, void 0);
    }
    else if (compare(ctxHealth.CurrentHealth, ctxHealth.LastHealth) < 0) {
        return new FSharpChoice$3(0, delta);
    }
    else {
        throw (new Error("Match failure"));
    }
}

export function $007CPercentLessHealth$007CPercentMoreHealth$007CPercentEqualHealth$007C(ctxHealth) {
    if (Math.abs(ctxHealth.CurrentHealthPercent - ctxHealth.LastHealthPercent) > 0.01) {
        if (compare(ctxHealth.CurrentHealthPercent, ctxHealth.LastHealthPercent) > 0) {
            return new FSharpChoice$3(1, void 0);
        }
        else if (compare(ctxHealth.CurrentHealthPercent, ctxHealth.LastHealthPercent) < 0) {
            return new FSharpChoice$3(0, void 0);
        }
        else {
            throw (new Error("Match failure"));
        }
    }
    else {
        return new FSharpChoice$3(2, void 0);
    }
}

export function $007CIsManaShieldEffect$007C_$007C(equal, objRef) {
    if (equal) {
        let matchValue;
        const self = skyrimPlatform.Actor;
        matchValue = skyrimPlatform.Actor.from(objRef);
        if (matchValue != null) {
            return matchValue;
        }
        else {
            return void 0;
        }
    }
    else {
        return void 0;
    }
}

export function $007CMagicEffectFromForm$007C_$007C(someForm) {
    if (someForm != null) {
        let matchValue;
        const self = skyrimPlatform.MagicEffect;
        matchValue = skyrimPlatform.MagicEffect.from(someForm);
        if (matchValue != null) {
            return matchValue;
        }
        else {
            return void 0;
        }
    }
    else {
        return void 0;
    }
}

export function $007CPerkFromForm$007C_$007C(someForm) {
    if (someForm != null) {
        let matchValue;
        const self = skyrimPlatform.Perk;
        matchValue = skyrimPlatform.Perk.from(someForm);
        if (matchValue != null) {
            return matchValue;
        }
        else {
            return void 0;
        }
    }
    else {
        return void 0;
    }
}

